src/generate.js

```diff
@@ -4,60 +4,94 @@ var self = module.exports;
 self.generate = function(options) {
-	// Set defaults.
-	if (!options.hasOwnProperty('length')) options.length = 10;
-	if (!options.hasOwnProperty('numbers')) options.numbers = false;
-	if (!options.hasOwnProperty('symbols')) options.symbols = false;
-	if (!options.hasOwnProperty('uppercase')) options.uppercase = true;
-	if (!options.hasOwnProperty('excludeSimilarCharacters')) options.excludeSimilarCharacters = false;
-
-	var lowercase = 'abcdefghijklmnopqrstuvwxyz',
-		uppercase = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ',
-		numbers = '0123456789',
-		symbols = '!@#$%^&*()+_-=}{[]|:;"/?.><,`~',
-		similarCharacters = /[ilLI|`oO0]/g;
-
-	// Generate character pool
-	var pool = lowercase;
-
-	// uppercase
-	if (options.uppercase) {
-		pool += uppercase;
-	}
-	// numbers
-	if (options.numbers) {
-		pool += numbers;
-	}
-	// symbols
-	if (options.symbols) {
-		pool += symbols;
-	}
-	// similar characters
-	if (options.excludeSimilarCharacters) {
-		pool = pool.replace(similarCharacters, '');
-	}
-
-	var password = '';
-	for (var i = 0; i < options.length; i++) {
-		password += pool[randomNumber(pool.length)];
-	}
-
-	return password;
+  // Set defaults.
+  if (!options.hasOwnProperty('length')) options.length = 10;
+  if (!options.hasOwnProperty('numbers')) options.numbers = false;
+  if (!options.hasOwnProperty('symbols')) options.symbols = false;
+  if (!options.hasOwnProperty('uppercase')) options.uppercase = true;
+  if (!options.hasOwnProperty('excludeSimilarCharacters')) options.excludeSimilarCharacters = false;
+  if (!options.hasOwnProperty('strict')) options.strict = false;
+
+  // Generate character pool
+  var pool = lowercase;
+
+  // uppercase
+  if (options.uppercase) {
+    pool += uppercase;
+  }
+  // numbers
+  if (options.numbers) {
+    pool += numbers;
+  }
+  // symbols
+  if (options.symbols) {
+    pool += symbols;
+  }
+  // similar characters
+  if (options.excludeSimilarCharacters) {
+    pool = pool.replace(similarCharacters, '');
+  }
+
+  var password = '';
+
+  if(options.strict){
+    var minStrictLength = 1 + (options.numbers ? 1 : 0) + (options.symbols ? 1 : 0) + (options.uppercase ? 1 : 0);
+    if(minStrictLength > options.length){
+      throw(new Error('Length should correlate with strict guidelines'));
+    } else {
+      var strictGroup = [];
+
+      strictGroup.push(lowercase[randomNumber(lowercase.length)]);
+
+      if (options.uppercase) {
+        strictGroup.push(uppercase[randomNumber(uppercase.length)]);
+      }
+      // numbers
+      if (options.numbers) {
+        strictGroup.push(numbers[randomNumber(numbers.length)]);
+      }
+      // symbols
+      if (options.symbols) {
+        strictGroup.push(symbols[randomNumber(symbols.length)]);
+      }
+
+      for (var j = minStrictLength; j < options.length; j++) {
+        strictGroup.push(pool[randomNumber(pool.length)]);
+      }
+
+      strictGroup.sort(function(){ return randomNumber(2) - 0.5;});
+      password = strictGroup.join('');
```

Comment:
The sort function only reorders the "strict groups", not the whole password, which isn't generated yet.

I wouldn't call it an edge case, but more of a possible attack vector. I really don't like the idea of adding any feature that knowingly decreases the entropy of a password when unnecessary. Password requirements like this are intended for humans, not machines, and although I'm ok with adding something to adhere to the requirements, I don't think the creator would think this is a good way of adhering to them. It's almost like, to require a number and a symbol, prefixing each password with `#0`.

You could arguably say the small degree of entropy loss is irrelevant. The thing is, I want this package to be grabbable by anyone on NPM and for them to have a reasonable expectation of perfect entropy — this would break that expectation.

I'd rather regenerate the password if it doesn't fit the requirements... does that work for you? Worst case then is a performance drop, which I'd prefer.